### ISSUE TITLE
[SDK] Shall we deprecate k8s sanitization when compiling pipeline yaml spec?

### ISSUE THREAD

**Issue by numerology** [kubeflow/pipelines#2629]
I feel like this is the root cause for issues like #2570 . I verified some samples and it looks like the pipeline works fine without sanitization as long as the parameter names are consistent with where they are declared.

Also it certainly prevents users from using complex typed parameters in the same (Pythonic) way they are defined. Thus I think the sanitization should be deprecated.

WDYT? @Ark-kun  @hongye-sun  @gaoning777 ?

**Comment by Ark-kun** [comment]
There are several aspects to this issue.
Argo has naming limitations for input/output, task and template names (plus maybe the placeholders and loop variable attribute keys). These need to be sanitized.

If that applies to the loop variables, then #2570 cannot be fully solved while using Argo. (It's similar to how in python you have limitations on keys when creating dictionaries using `dict(a=1, key with spaces=2)`).

P.S. Sanitation function can be relaxed to enable underscored names (enabled by Argo v2.3.0).

**Comment by numerology** [comment]
@Ark-kun 
'key with spaces' should not work for sure, but I checked it seems like capital letter and underscore should work just fine.

Let me rephrase, shall we refactor/loosen the sanitization to keep the Pythonic naming convention (or something bit more)?

**Comment by Ark-kun** [comment]
>Let me rephrase, shall we refactor/loosen the sanitization

Yes. DSL-Compiler should only sanitize the names to meet Argo requirements: https://github.com/argoproj/argo/blob/3bead0db3d2777638992ba5e11a2de1c65be162c/workflow/validate/validate.go#L1028

`^[-a-zA-Z0-9_]+[-a-zA-Z0-9_]*$`
